clang/lib/CodeGen/CGCilk.cpp

```diff
@@ -471,22 +471,24 @@ void CodeGenFunction::EmitCilkScopeStmt(const CilkScopeStmt &S) {
     if (ThisScopeIsOutermost && !CurSyncRegion) {
       llvm::Instruction *TapirRTStart = Builder.CreateCall(
           CGM.getIntrinsic(llvm::Intrinsic::tapir_runtime_start));
-      // Mark the end of the _Cilk_scope with tapir_runtime_end.
+      // Mark the end of the cilk_scope with tapir_runtime_end.
       EHStack.pushCleanup<TapirRuntimeEndCleanup>(NormalAndEHCleanup,
                                                   TapirRTStart);
     }
     // Create a nested synced scope.
-    SyncRegionRAII SyncReg(*this);
-    bool BodyIsCompoundStmt = isa<CompoundStmt>(S.getBody());
-    if (BodyIsCompoundStmt)
+    if (isa<CompoundStmt>(S.getBody())) {
+      SyncRegionRAII SyncReg(*this);
       ScopeIsSynced = true;
-
-    // Emit the spawned statement.
-    EmitStmt(S.getBody());
+      EmitStmt(S.getBody());
+    } else {
+      PushSyncRegion()->addImplicitSync();
```

Comment:
It took me a while to figure out why these changes were necessary, given that the `SyncRegionRAII` is meant to handle pushing and popping of sync regions.

I think the simpler approach to resolving this issue is to set `ScopeIsSynced` to `true` before pushing the `TapirRuntimeEndCleanup`, e.g., on line 472.  That flag is used to ensure that implicit syncs and cleanups/destructors are ordered correctly within a lexical scope.  Does that approach work for your tests?